### PR TITLE
Dem refactor distribution

### PR DESCRIPTION
**Pull request by OGaboriault**
# Description of the problem

- Using many particle types with different size distribution wasn't possible. 
- The maximum_particle_size calculation was about to become false if we add new type of distribution. 

# Description of the solution

- New way to parse the size_distribution. 
- Distribution class have two new virtual function. 
- Insertion class stores a vector of Distribution. 
- find_maximum_particle_size  class has been removed. 

# How Has This Been Tested?

- All test are still pass.

# Future changes

- Add new type of distribution.

# Comments

- Probably a lot of stuff to change with this PR.

### CODE REVIEW

**Comment by blaisb** [comment]
This HAKJSHLKAHSLK periodic test will need to be fixed. I'll try to fix it this week. Pisses me off to no avail :)

**Comment by OGaboriault** [comment]
Ok, I went from 2k modified lines to 375 lines with this last push. I think this will be help you with your reviews.

File: source/core/parameters_lagrangian.cc [hunk 1]
```diff
@@ -175,70 +172,71 @@ namespace Parameters
...
+      g[0] = prm.get_double("gx");
+      g[1] = prm.get_double("gy");
+      g[2] = prm.get_double("gz");
+
+      particle_type_number = prm.get_integer("number of particle types");
 
-        if (particle_type_number >= 1)
+      if (particle_type_number >= 1)
+        {
+          prm.enter_subsection("particle type 0");
```
Comment (blaisb) [inline source/core/parameters_lagrangian.cc:184]: Instead of having a ton of ifs like that, why not just make a single for loop?

**Review by blaisb — CHANGES_REQUESTED** [review]
First comment, will finish tonight

File: source/dem/dem.cc [hunk 2]
```diff
@@ -203,26 +204,59 @@ DEMSolver<dim>::DEMSolver(DEMSolverParameters<dim> dem_parameters)
         parameters.model_parameters.solid_fraction_threshold);
     }
 
+  distribution_object_container.reserve(
+    parameters.lagrangian_physical_properties.particle_type_number);
+  maximum_particle_diameter = 0;
+  for (unsigned int counter = 0;
```
Comment (voferreira) [inline source/dem/dem.cc:210]: "counter" doesn't feel like an iterator. I would use "type_number".

File: source/dem/dem.cc [hunk 3]
```diff
@@ -203,26 +204,59 @@ DEMSolver<dim>::DEMSolver(DEMSolverParameters<dim> dem_parameters)
         parameters.model_parameters.solid_fraction_threshold);
     }
 
+  distribution_object_container.reserve(
```
Comment (voferreira) [inline source/dem/dem.cc:207]: Given line 74, do you need to reserve?

File: source/dem/dem.cc [hunk 4]
```diff
@@ -203,26 +204,59 @@ DEMSolver<dim>::DEMSolver(DEMSolverParameters<dim> dem_parameters)
...
+
+
   // Calling input_parameter_inspection to evaluate input parameters in the
   // parameter handler file, finding maximum particle diameter used in
   // polydisperse systems
-  maximum_particle_diameter =
-    find_maximum_particle_size(parameters.lagrangian_physical_properties,
-                               standard_deviation_multiplier);
+
+
```
Comment (voferreira) [inline source/dem/dem.cc:243]: Extra line?

File: source/fem-dem/cfd_dem_coupling.cc [hunk 5]
```diff
@@ -1313,19 +1310,45 @@ CFDDEMSolver<dim>::dem_setup_contact_parameters()
...
+  distribution_object_container.reserve(
+    dem_parameters.lagrangian_physical_properties.particle_type_number);
+
+  maximum_particle_diameter = 0.;
+  for (unsigned int counter = 0;
+       counter <
+       dem_parameters.lagrangian_physical_properties.particle_type_number;
+       counter++)
+    {
+      if (dem_parameters.lagrangian_physical_properties.distribution_type.at(
```
Comment (voferreira) [inline source/fem-dem/cfd_dem_coupling.cc:1322]: Same comment about the variable

**Review by voferreira — CHANGES_REQUESTED** [review]
Good Job. It is good to have distribution types separately like this. Just some small comments here and there but nothing fancy.

File: include/fem-dem/cfd_dem_coupling.h [hunk 6]
```diff
@@ -272,6 +271,7 @@ class CFDDEMSolver : public GLSVANSSolver<dim>
   Visualization<dim>            visualization_object;
   BoundaryCellsInformation<dim> boundary_cell_object;
 
+
```
[the remark this answers is not in the document]
Comment (OGaboriault) [reply include/fem-dem/cfd_dem_coupling.h:274]: I have add something and removed it afterward... sorry

File: include/dem/input_parameter_inspection.h [hunk 7]
```diff
@@ -31,16 +32,15 @@ using namespace std;
  *
  * @param dem_parameters Input DEM parameters in the parameter handler file
  * @param pcout Printing in parallel
- * @param standard_deviation_multiplier Constant standard deviation multiplier.
- * It is defined in the dem constructor and it is equal to 2.5 to cover 99% of
- * the particle size distribution
+ * @param distribution_object_container Contain all the types of distribution being used for each type of particle.
```
[the remark this answers is not in the document]
Comment (OGaboriault) [reply include/dem/input_parameter_inspection.h:35]: You are right, but clang-format doesn't do anything...

I will put this on two lines.

File: source/core/parameters_lagrangian.cc (hunk 1 above)
Comment (OGaboriault) [reply source/core/parameters_lagrangian.cc:184]: I will try to do something similar to the parsing of the solid object.

Comment (OGaboriault) [reply source/core/parameters_lagrangian.cc]: Done

**Comment by blaisb** [comment]
@OGaboriault can you rebase?

**Comment by OGaboriault** [comment]
Done

**Review by voferreira — APPROVED** [review]

**Pull request merged**